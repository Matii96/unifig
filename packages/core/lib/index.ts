export * from './adapters';
export * from './manager';
export * from './properties-mapping';
export * from './validator';
