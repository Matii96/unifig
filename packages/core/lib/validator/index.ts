export * from './decorators';
export * from './errors';
