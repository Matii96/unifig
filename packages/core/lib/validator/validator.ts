import { ConfigValidationError } from './errors/validation.error';

export interface Validator {
  validate(configs: object[]): ConfigValidationError | undefined;
}
