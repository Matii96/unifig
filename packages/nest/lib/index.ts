export * from './injection';
export * from './config.module';
