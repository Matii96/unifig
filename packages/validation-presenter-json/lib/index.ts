export * from './to-json.options';
export * from './to-json';
