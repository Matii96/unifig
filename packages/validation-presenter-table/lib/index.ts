export * from './to-table.options';
export * from './to-table';
