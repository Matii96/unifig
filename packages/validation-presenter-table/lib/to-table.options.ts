export interface ToTableOptions {
  /**
   * @default norc
   */
  border?: 'honeywell' | 'norc' | 'ramac' | 'void';
}
