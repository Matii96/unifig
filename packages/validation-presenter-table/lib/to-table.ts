import {
  ConfigTemplateValidationError,
  ConfigValidationError,
  ConfigPropertyValidationError,
  ConfigSubtemplateValidationError,
} from '@unifig/core';
import { getBorderCharacters, SpanningCellConfig, table } from 'table';
import { TemplateRow } from './template-row/template-row';
import { ToTableOptions } from './to-table.options';

const HEADER = ['Template', 'Property', 'Type', 'Source', 'Current Value', 'Failed constraints'];

export const toTable = (
  validationException: ConfigValidationError,
  options: ToTableOptions = {}
): string => {
  const tableRows = formatTableRowsGroup(validationException.errors);
  const tableData = [HEADER, ...tableRows.flatMap(({ templateTableData }) => templateTableData)];

  return table(tableData, {
    columns: [{ alignment: 'left', width: 16 }],
    spanningCells: tableRows.map(({ spanningCells }) => spanningCells),
    border: getBorderCharacters(options.border ?? 'norc'),
  });
};

const formatTableRowsGroup = (failedValidations: ConfigTemplateValidationError[]) => {
  let spanningCellsRowIdx = 1;
  return failedValidations.map((failedValidation) => {
    const templateTableData = failedValidation.errors
      .flatMap((error) => formatTemplateRows(error))
      .map((row) => row.toArray());
    templateTableData[0][0] = failedValidation.template.name;

    const spanningCells: SpanningCellConfig = {
      col: 0,
      row: spanningCellsRowIdx,
      rowSpan: templateTableData.length,
      verticalAlignment: 'middle',
    };
    spanningCellsRowIdx += templateTableData.length;
    return { templateTableData, spanningCells };
  });
};

const formatTemplateRows = (
  error: ConfigPropertyValidationError | ConfigSubtemplateValidationError,
  parentPrefix = ''
): TemplateRow[] => {
  if (error instanceof ConfigPropertyValidationError) {
    return [TemplateRow.fromValidationError(error, parentPrefix)];
  }
  if (error instanceof ConfigSubtemplateValidationError) {
    const propertyParentPrefix = parentPrefix + error.property + '.';
    const subtemplateRows = error.children.flatMap((child) =>
      formatTemplateRows(child, propertyParentPrefix)
    );
    return error.failedConstraints
      ? [TemplateRow.fromValidationError(error), ...subtemplateRows]
      : subtemplateRows;
  }
  throw new Error('Passed plain validation object: ' + JSON.stringify(error));
};
