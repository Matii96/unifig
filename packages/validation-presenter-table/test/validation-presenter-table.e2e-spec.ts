import { ConfigValidationError } from '@unifig/core';
import { mockFailed1dValidation, mockFailed2dValidation } from '../lib/validator.mocks';
import { toTable } from '../lib';

describe('@unifig/validation-presenter-table (e2e)', () => {
  it('should format table from 1d error', () => {
    expect(toTable(new ConfigValidationError([mockFailed1dValidation()]))).toMatchSnapshot();
  });

  it('should format table from 2d errors', () => {
    expect(toTable(new ConfigValidationError(mockFailed2dValidation()))).toMatchSnapshot();
  });
});
