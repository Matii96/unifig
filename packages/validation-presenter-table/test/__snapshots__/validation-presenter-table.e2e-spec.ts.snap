// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@unifig/validation-presenter-table (e2e) should format table from validation errors 1`] = `
"┌──────────────────┬─────────────┬────────┬─────────────┬───────────────┬────────────────────┐
│ Template         │ Property    │ Type   │ Source      │ Current Value │ Failed constraints │
├──────────────────┼─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│                  │ port        │ Number │ PORT        │ "not-a-port"  │ isInt              │
│                  ├─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│ TemplateMock     │ db.url      │ String │ DB_URL      │ undefined     │ isString           │
│                  ├─────────────┼────────┼─────────────┼───────────────┼────────────────────┤
│                  │ db.password │ String │ DB_PASSWORD │ undefined     │ isString           │
└──────────────────┴─────────────┴────────┴─────────────┴───────────────┴────────────────────┘
"
`;
